bpo-35424: emit ResourceWarning at multiprocessing.Pool destruction

vstinner opened the pull request:
If a multiprocessing.Pool is not joined explicitly, it now emits a
ResourceWarning warning.

* Add INIT state
* Pool state now initialized to INIT and only set to RUN when
  Pool.__init__() completes successfully.
* Initialize some attributes earlier int Pool.__init__() to ensure
  that they exist in Pool.__del__()

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-35424](https://bugs.python.org/issue35424) -->
https://bugs.python.org/issue35424
<!-- /issue-number -->

pablogsal commented on Lib/multiprocessing/pool.py:
```diff
@@ -219,6 +232,23 @@ def __init__(self, processes=None, initializer=None, initargs=(),
+            _warn(f"unclosed running multiprocessing pool {self!r}", ResourceWarning,
+                  source=self)
+        elif self._state != INIT and not self._joined:
+            _warn(f"multiprocessing pool {self!r} was not joined",
```
Should this be `stopped` instead of `joined`?

vstinner replied:
I don't know how to describe the state here... close() or terminate() has been called. "was not joined" means that .join() should be called and it has not been called. 

"stopped" implies that there is stop() method, no? But there is no stop() method.

pablogsal replied:
Sorry, I meant `close()` instead of `stop()`

vstinner replied:
close() doesn't ensure that all resources are released. Currently, join() must be called to ensure that all processes and threads completed.

pablogsal replied:
👍

eamanu replied:
I thing close is more clearly, isn't?

vstinner replied:
Read previous comments.

pitrou replied:
As long as the user has called `terminate` or `close` (or used a context manager), a warning shouldn't be emitted.

pablogsal commented on Lib/multiprocessing/pool.py:
```diff
@@ -219,6 +232,23 @@ def __init__(self, processes=None, initializer=None, initargs=(),
+        state = _STATES.get(self._state, self._state)
+        return (f'<{cls.__module__}.{cls.__qualname__} '
+               f'state={state} '
+               f'pool#={len(self._pool)}>')
```
Is this "#" on purpose?

vstinner replied:
It's on purpose. '#' means 'size' or 'count': it's the number of items (processes or threads) in the pool. It's not the right syntax?

pablogsal replied:
I understand now. I think I was confused about searching for some other meaning for the '#'.

vstinner replied:
I can remove '#' if you prefer, I don't really care :-) Or rename to 'pool_size' maybe?

pablogsal replied:
I am happy with the '#' :)

pitrou replied:
I'd rather have "pool size".

tzickel commented on Lib/multiprocessing/pool.py:
```diff
@@ -219,6 +232,23 @@ def __init__(self, processes=None, initializer=None, initargs=(),
             )
+        self._state = RUN
+
+    def __del__(self, _warn=warnings.warn):
```
it's possible to call del with extra arguments ?

vstinner replied:
No, you cannot. That's a trick to keep warnings.warn() alive during Python finalization. See other __del__() method of subprocess and asyncio module for example.

pitrou reviewed (changes requested):
I'd like to see the semantics changed a bit. Legitimate uses shouldn't see a warning. Also I notice you didn't add any tests.

vstinner committed:
bpo-35424: multiprocessing.Pool emits ResourceWarning

multiprocessing.Pool destructor now emits ResourceWarning
if the pool is still running.

pitrou commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -2577,6 +2577,22 @@ def test_enter(self):
+        pool.terminate()
+        pool.join()
+
+        # force state to RUN to emit ResourceWarning in __del__()
```
Yuck. Why can't you test the actual situation? (start a pool then delete it)

vstinner replied:
If you don't cleanup properly a pool, bad things happens. You can leak processes and threads and it's not easy to write a test which wait properly until they complete. If the test doesn't wait until they complete, the test can be marked as ENV_CHANGED and the whole test suite fails on buildbots. Leaking threads/processes have side effects on following tests which cause random failures. There is a "coverage" job on Travis CI which runs the full test suite sequentially for example.

pitrou replied:
Well, there was a PR to clean up pools properly, but AFAIR you reverted it...

vstinner replied:
I reverted https://bugs.python.org/issue34172 change because it introduced a backward incompatible change in stable branches. I also reverted the change in the master branch to open a more general discussion: https://mail.python.org/pipermail/python-dev/2018-December/155946.html

The reverted change isn't magic: it doesn't fix all issues. Even with this fix, the destructor still doesn't join many threads and so regrtest will still complain. The .join() method has to be called explicitly to really make sure that a test doesn't leak resources.

The reverted change had a test on "del pool" and this test caused issues on buildbots because it leaked resources (processes, threads). IMHO testing that the destructor calls terminate() deserves its own separated test.

In this PR, the test is restricted to check that the destructor emits a ResourceWarning.

Moreover, to come back to https://bugs.python.org/issue34172 maybe you haven't noticed, by this PR should also help https://bugs.python.org/issue34172 because the problematic example ("for x in multiprocessing.Pool().imap(int, ["4", "3"]): print(x)") should now emit a ResourceWarning (once the reference cycle will be broken again). It's better than nothing to warn developers that something is wrong :-) See also https://bugs.python.org/issue35378

... Sorry for the disgression, but it seems like you asked for it :-)

vstinner replied:
Do you want me to change the test, or is it ok written like that, with the explanation in my previous comment?

pitrou replied:
Fair enough, let's keep it like that :-)

pitrou commented on Lib/multiprocessing/pool.py:
```diff
@@ -216,6 +221,12 @@ def __init__(self, processes=None, initializer=None, initargs=(),
+        self._state = RUN
+
+    def __del__(self, _warn=warnings.warn):
+        if self._state == RUN:
```
Shouldn't you freeze `RUN` in the `__del__` parameters as well?

vstinner replied:
done

vstinner committed:
__del__() copies also RUN global
